### ISSUE TITLE
Can xgboost use custom loss functions that require tracking additional parameters?

### ISSUE THREAD

**Issue by jeffwong** [dmlc/xgboost#513]
I would like to implement a cox loss function for xgboost. Cox has a likelihood function which can be borrowed from the gbm package, but it also requires keeping track of the censor flag.  In this example

    logregobj <- function(preds, dtrain) {
      labels <- getinfo(dtrain, "label")
      preds <- 1/(1 + exp(-preds))
      grad <- preds - labels
      hess <- preds * (1 - preds)
      return(list(grad = grad, hess = hess))
    }

we are only keeping track of labels and predictions. Do all loss functions have to have this form? Is there a way to extend it to keep track of misc vectors too?

**Comment by BlindApe** [comment]
You can pass as attribute:

attr(dtrain, 'censor') <- censor_flag
logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  censor <- attr(dtrain, 'censor') 
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  return(list(grad = grad, hess = hess))
}

Now you get censor inside custom eval function (works with cv folds too).

**Comment by walterreade** [comment]
@BlindApe - Great tip. Thx.

Along these lines, is anyone aware of where I can find an example of using XGBoost for survival analysis?

**Comment by BrianMiner** [comment]
@jeffwong did you ever implement and curious if you could share?

**Comment by mpearmain** [comment]
@walterreade Did you ever find an XGB example for survival analysis?

**Comment by walterreade** [comment]
@mpearmain - Nope. Still on my list to find / figure out.
